File: pipeline/urlstate.py
"""The permalink codec: `PivotQuery` <-> query string. CI-ONLY reference implementation.

pipeline/ never runs in prod -- the server does this in TypeScript. This module IS the
spec M3b's port is verified against, same relationship as pivot.py's `render_pivot` to the
templates it renders.

**The encoding is a frozen public contract from the first shipped link.** Keys are short and
readable, not a base64 JSON blob: this audience hand-edits permalinks and pastes them into
forums, where an opaque blob is both unreadable and un-editable. `v=1` exists so a future
incompatible change can migrate a link instead of silently misreading it.

**Decode is TOTAL.** An unknown query-string key, an unrecognised or missing `v`, or anything
that fails `render_pivot`'s own allowlist/structural validation (unknown dimension/measure/
sort key, unknown grain/grouping, empty dimension or measure list, a non-positive limit, a
filter with no values) is a rejection via `UrlStateError` -- never a silent drop to a
default. A permalink that quietly renders a *different* query than the one it encodes still
screenshots as authoritative, which is worse than one that errors.

**Division of validation labour** (per the project rule against a second, drifting
validator): identifier and structural validation -- is this dimension/measure/sort key on
the allowlist, is the grain/grouping recognised, is the limit a positive int, does every
filter have values -- is reused *as-is* from `pipeline.pivot.render_pivot`, which this module
calls purely to validate a candidate `PivotQuery` (its rendered SQL/params are discarded).
That is the ONE place those rules live.

What `render_pivot` cannot check, because it is URL syntax rather than a `PivotQuery` field,
is validated HERE instead:
  - the `v` key itself (present, and equal to `URL_VERSION`);
  - unknown query-string keys (anything not in `_ALLOWED_KEYS`);
  - the shape of `t` (`YYYY-MM:YYYY-MM`) and `f` (`key:val1,val2,...`) tokens;
  - parsing `n` as an integer at all (a non-numeric limit is a codec-level parse failure,
    not a value `render_pivot` would ever see as anything but the wrong Python type).
"""

from __future__ import annotations

from urllib.parse import parse_qsl

from pipeline.pivot import GRAINS, GROUPINGS, PivotError, PivotQuery, render_pivot

URL_VERSION = 1

#: grain <-> its short URL token. Frozen alongside the rest of the key schema.
_GRAIN_TO_URL = {"segment": "seg", "route": "route"}
_URL_TO_GRAIN = {v: k for k, v in _GRAIN_TO_URL.items()}

#: grouping <-> its short URL token.
_GROUPING_TO_URL = {"operating": "op", "mainline": "ml"}
_URL_TO_GROUPING = {v: k for k, v in _GROUPING_TO_URL.items()}

#: The complete key vocabulary. Anything else on the query string is a rejection, not a
#: silently-ignored extra -- see module docstring.
_ALLOWED_KEYS = frozenset({"v", "k", "d", "m", "t", "f", "s", "n", "g"})

assert set(_GRAIN_TO_URL) == GRAINS, "grain URL-token map has drifted from pivot.GRAINS"
assert set(_GROUPING_TO_URL) == GROUPINGS, "grouping URL-token map has drifted from pivot.GROUPINGS"


class UrlStateError(Exception):
    """A query string could not be decoded into a valid `PivotQuery`."""


def encode(q: PivotQuery) -> str:
    """Serialize a `PivotQuery` to a query string (no leading `?`).

    Deterministic and total over every `PivotQuery` the caller can construct: every field
    always has a key, in a fixed order, so two equal queries always encode to the same
    string -- required for the `url -> state -> url` round trip.
    """
    parts = [
        f"v={URL_VERSION}",
        f"k={_GRAIN_TO_URL[q.grain]}",
        f"d={','.join(q.dimensions)}",
        f"m={','.join(q.measures)}",
        f"t={q.time_from}:{q.time_to}",
    ]
    for key, values in q.filters:
        parts.append(f"f={key}:{','.join(values)}")
    if q.sort is not None:
        prefix = "-" if q.sort_desc else ""
        parts.append(f"s={prefix}{q.sort}")
    parts.append(f"n={q.limit}")
    parts.append(f"g={_GROUPING_TO_URL[q.grouping]}")
    return "&".join(parts)


def _parse_time_range(raw: str) -> tuple[str, str]:
    if raw.count(":") != 1:
        raise UrlStateError(f"malformed time range {raw!r}, expected 'YYYY-MM:YYYY-MM'")
    time_from, time_to = raw.split(":")
    if not time_from or not time_to:
        raise UrlStateError(f"malformed time range {raw!r}, expected 'YYYY-MM:YYYY-MM'")
    return time_from, time_to


def _parse_filter(raw: str) -> tuple[str, tuple[str, ...]]:
    if ":" not in raw:
        raise UrlStateError(f"malformed filter {raw!r}, expected 'key:val1,val2,...'")
    key, values_raw = raw.split(":", 1)
    values = tuple(v for v in values_raw.split(",") if v)
    if not key or not values:
        raise UrlStateError(f"malformed filter {raw!r}, expected 'key:val1,val2,...'")
    return key, values


def decode(qs: str, con) -> PivotQuery:
    """Parse a query string into a `PivotQuery`, validated via `con` against the catalog.

    Raises `UrlStateError` for anything malformed at the URL-syntax level (see module
    docstring), and re-raises `pipeline.pivot.PivotError` -- from `render_pivot`, called
    purely to validate -- as `UrlStateError` so callers only need to catch one exception
    type from this module.
    """
    pairs = parse_qsl(qs, keep_blank_values=True)

    unknown = [key for key, _ in pairs if key not in _ALLOWED_KEYS]
    if unknown:
        raise UrlStateError(f"unknown query key(s): {sorted(set(unknown))}")

    single: dict[str, str] = {}
    filters_raw: list[str] = []
    for key, value in pairs:
        if key == "f":
            filters_raw.append(value)
        else:
            # Last one wins for a repeated non-'f' key; nothing in the schema needs the
            # first, and 'f' is the only key documented as repeatable.
            single[key] = value

    version_raw = single.get("v")
    if version_raw is None:
        raise UrlStateError("missing url version ('v')")
    try:
        version = int(version_raw)
    except ValueError:
        raise UrlStateError(f"unrecognised url version {version_raw!r}") from None
    if version != URL_VERSION:
        raise UrlStateError(f"unrecognised url version {version!r}, expected {URL_VERSION}")

    grain_token = single.get("k", "")
    grain = _URL_TO_GRAIN.get(grain_token, grain_token)

    dimensions = tuple(single["d"].split(",")) if single.get("d") else ()
    measures = tuple(single["m"].split(",")) if single.get("m") else ()

    if "t" not in single:
        raise UrlStateError("missing time range ('t')")
    time_from, time_to = _parse_time_range(single["t"])

    filters = tuple(_parse_filter(raw) for raw in filters_raw)

    sort: str | None = None
    sort_desc = True
    if "s" in single:
        raw_sort = single["s"]
        if raw_sort.startswith("-"):
            sort, sort_desc = raw_sort[1:], True
        else:
            sort, sort_desc = raw_sort, False

    if "n" in single:
        try:
            limit = int(single["n"])
        except ValueError:
            raise UrlStateError(f"limit ('n') must be an integer, got {single['n']!r}") from None
    else:
        limit = 100

    grouping_token = single.get("g", "operating")
    grouping = _URL_TO_GROUPING.get(grouping_token, grouping_token)

    candidate = PivotQuery(
        grain=grain,
        dimensions=dimensions,
        measures=measures,
        time_from=time_from,
        time_to=time_to,
        filters=filters,
        sort=sort,
        sort_desc=sort_desc,
        limit=limit,
        grouping=grouping,
    )

    try:
        render_pivot(candidate, con)
    except PivotError as exc:
        raise UrlStateError(str(exc)) from exc

    return candidate

File: app/src/lib/pivot/urlstate.ts
import { PivotError, normalizeQuery, type Grain, type Grouping, type PivotQuery } from "@/lib/pivot/types";
import { renderPivot } from "@/lib/pivot/render";
import type { Allowlist } from "@/lib/pivot/allowlist";

export const URL_VERSION = 1;

export class UrlStateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UrlStateError";
  }
}

const MONTH_RE = /^\d{4}-(0[1-9]|1[0-2])$/;
const GRAIN_TO_URL: Record<Grain, string> = { segment: "seg", route: "route" };
const URL_TO_GRAIN: Record<string, Grain> = { seg: "segment", route: "route" };
const GROUPING_TO_URL: Record<Grouping, string> = { operating: "op", mainline: "ml" };
const URL_TO_GROUPING: Record<string, Grouping> = { op: "operating", ml: "mainline" };
const ALLOWED_KEYS = new Set(["v", "k", "d", "m", "t", "f", "s", "n", "g"]);

/** Python's urllib.parse.quote(v, safe=""). encodeURIComponent leaves ! * ' ( ) literal;
 * Python does not, and the goldens pin Python. Real data hits this: 119 carrier codes
 * carry BTS's parenthesised suffix. */
export function quote(s: string): string {
  return encodeURIComponent(s).replace(
    /[!'()*]/g,
    (c) => "%" + c.charCodeAt(0).toString(16).toUpperCase(),
  );
}

export function encode(q: PivotQuery): string {
  const parts = [
    `v=${URL_VERSION}`,
    `k=${GRAIN_TO_URL[q.grain]}`,
    `d=${q.dimensions.join(",")}`,
    `m=${q.measures.join(",")}`,
    `t=${q.timeFrom}:${q.timeTo}`,
  ];
  for (const [key, values] of q.filters) {
    parts.push(`f=${quote(key)}:${values.map(quote).join(",")}`);
  }
  if (q.sort !== null) parts.push(`s=${q.sortDesc ? "-" : ""}${q.sort}`);
  parts.push(`n=${q.limit}`);
  parts.push(`g=${GROUPING_TO_URL[q.grouping]}`);
  return parts.join("&");
}

/** Split on the literal & and the first =. Deliberately NOT URLSearchParams: decoding must
 * happen only after every structural delimiter -- including f's own : and , -- has done its
 * job, or a percent-encoded structural comma is silently corrupted. */
function splitPairs(qs: string): [string, string][] {
  if (!qs) return [];
  const out: [string, string][] = [];
  for (const chunk of qs.split("&")) {
    if (!chunk) continue;
    const i = chunk.indexOf("=");
    out.push(i === -1 ? [chunk, ""] : [chunk.slice(0, i), chunk.slice(i + 1)]);
  }
  return out;
}

function parseFilter(raw: string): [string, string[]] {
  const i = raw.indexOf(":");
  if (i === -1) throw new UrlStateError(`malformed filter '${raw}', expected 'key:val1,val2,...'`);
  const key = decodeURIComponent(raw.slice(0, i));
  const values = raw.slice(i + 1).split(",").filter((v) => v).map(decodeURIComponent);
  if (!key || values.length === 0) {
    throw new UrlStateError(`malformed filter '${raw}', expected 'key:val1,val2,...'`);
  }
  return [key, values];
}

export function decode(qs: string, a: Allowlist): PivotQuery {
  const seen = new Set<string>();
  const filters: [string, string[]][] = [];
  const single: Record<string, string> = {};

  for (const [key, raw] of splitPairs(qs)) {
    if (!ALLOWED_KEYS.has(key)) {
      throw new UrlStateError(
        `unknown query key '${key}': allowed keys are ${[...ALLOWED_KEYS].sort().join(", ")}`,
      );
    }
    if (key === "f") {
      filters.push(parseFilter(raw));
      continue;
    }
    if (seen.has(key)) {
      throw new UrlStateError(
        `duplicate key '${key}': a permalink that quietly renders a different query than it ` +
          "encodes is worse than one that errors",
      );
    }
    seen.add(key);
    single[key] = decodeURIComponent(raw);
  }

  for (const required of ["v", "k", "d", "m", "t", "n", "g"]) {
    if (!(required in single)) throw new UrlStateError(`missing required key '${required}'`);
  }
  if (single.v !== String(URL_VERSION)) {
    throw new UrlStateError(`unsupported version '${single.v}', expected ${URL_VERSION}`);
  }

  const grain = URL_TO_GRAIN[single.k];
  if (!grain) throw new UrlStateError(`unknown grain token '${single.k}'`);
  const grouping = URL_TO_GROUPING[single.g];
  if (!grouping) throw new UrlStateError(`unknown grouping token '${single.g}'`);

  const colon = single.t.indexOf(":");
  if (colon === -1 || single.t.indexOf(":", colon + 1) !== -1) {
    throw new UrlStateError(`malformed time range '${single.t}', expected 'YYYY-MM:YYYY-MM'`);
  }
  const timeFrom = single.t.slice(0, colon);
  const timeTo = single.t.slice(colon + 1);
  if (!MONTH_RE.test(timeFrom) || !MONTH_RE.test(timeTo)) {
    throw new UrlStateError(`malformed time range '${single.t}', expected 'YYYY-MM:YYYY-MM'`);
  }

  if (!/^\d+$/.test(single.n)) throw new UrlStateError(`limit must be a positive integer, got '${single.n}'`);
  const limit = Number(single.n);

  let sort: string | null = null;
  let sortDesc = true;
  if ("s" in single) {
    sortDesc = single.s.startsWith("-");
    sort = sortDesc ? single.s.slice(1) : single.s;
    if (!sort) throw new UrlStateError("malformed sort key");
  }

  const q = normalizeQuery({
    grain,
    dimensions: single.d.split(","),
    measures: single.m.split(","),
    timeFrom,
    timeTo,
    filters,
    sort,
    sortDesc,
    limit,
    grouping,
  });

  // Identifier and structural validation is reused as-is -- one allowlist, not two.
  try {
    renderPivot(q, a);
  } catch (e) {
    if (e instanceof PivotError) throw new UrlStateError(e.message);
    throw e;
  }
  return q;
}
